staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,615 @@
+// are non-empty queues and the number currently executing is less than the
+// assured concurrency value.  The body of the loop uses the fair queuing
+// technique to pick a queue, dequeue the request at the head of that
+// queue, increment the count of the number executing, and send true
+// to the request's channel.
+func (qs *queueSetImpl) dequeueWithChannelAsMuchAsPossible() {
+	for !qs.isEmpty() && qs.getRequestsExecutingLocked() < qs.concurrencyLimit {
+		_, ok := qs.dequeueWithChannel()
+		// TODO(aaron-prindle) verify checking ok makes senes
+		if !ok {
+			break
+		}
+	}
+}
+
+// dequeueWithChannel is convenience method for dequeueing packets that
+// require a message to be sent through the packets channel
+// this is a required pattern for the QueueSet the queueSetImpl supports
+func (qs *queueSetImpl) dequeueWithChannel() (*fq.Request, bool) {
+	req, ok := qs.dequeue()
+	if !ok {
+		return nil, false
+	}
+	qs.counter.Add(1)
+	req.DequeueChannel <- true
+	close(req.DequeueChannel)
+	return req, ok
+}
+
+// removeIndex uses reslicing to remove an index from a slice
+func removeQueue(s []*fq.Queue, index int) []*fq.Queue {
+	return append(s[:index], s[index+1:]...)
+}
+
+// removeQueueAndUpdateIndexes uses reslicing to remove an index from a slice
+// and then updates the 'Index' field of the queues to be correct
+func removeQueueAndUpdateIndexes(queues []*fq.Queue, index int) []*fq.Queue {
+	removedQueues := removeQueue(queues, index)
+	for i := index; i < len(removedQueues); i++ {
+		removedQueues[i].Index--
+	}
+	return removedQueues
+}
+
+// FinishRequest is a callback that should be used when a previously dequeued packet
+// has completed it's service.  This callback updates imporatnt state in the
+// queueSetImpl
+func (qs *queueSetImpl) finishRequest(r *fq.Request) {
+
+	S := qs.clock.Since(r.StartTime).Seconds()
+
+	// When a request finishes being served, and the actual service time was S,
+	// the queue’s virtual start time is decremented by G - S.
+	r.Queue.VirtualStart -= qs.estimatedServiceTime - S
+
+	// request has finished, remove from requests executing
+	r.Queue.RequestsExecuting--
+
+	// Logic to remove quiesced queues
+	// >= as QueueIdx=25 is out of bounds for desiredNumQueues=25 [0...24]
+	if r.Queue.Index >= qs.desiredNumQueues &&
+		len(r.Queue.Requests) == 0 &&
+		r.Queue.RequestsExecuting == 0 {
+		qs.queues = removeQueueAndUpdateIndexes(qs.queues, r.Queue.Index)
+
+		// decrement here to maintain the invariant that (qs.robinIdx+1) % numQueues
+		// is the index of the next queue after the one last dispatched from
+		if qs.robinIdx >= -r.Queue.Index {
+			qs.robinIdx--
+		}
+
+		// At this point, if the qs is quiescing,
+		// has zero requests executing, and has zero requests enqueued
+		// then a call to the EmptyHandler should be forked.
+		if qs.emptyHandler != nil && qs.numRequestsEnqueued == 0 &&
+			qs.getRequestsExecutingLocked() == 0 {
+			qs.counter.Add(1)
+			go qs.emptyHandler.HandleEmpty()
```

Comment:
As noted at the other call to the EmptyHandler, the responsibility to call `qs.counter.Add(-1)` should be here.
Also, there is a concurrency bug: reading the `emptyHandler` field of `qs` without holding the lock.
We might fix both as follows.

> go func(eh fq.EmptyHandler) {eh.HandleEmpty(); qs.counter.Add(-1)} (qs.emptyHandler)

Reply:
Ah makes sense, I wasn't sure if that would be handled by HandleEmpty but this makes it more clear.  I've used the suggested implementation:
```
		go func(eh fq.EmptyHandler) {
			eh.HandleEmpty()
			qs.counter.Add(-1)
		}(qs.emptyHandler)
```